File: packages/runar-java/src/test/java/runar/lang/sdk/IntentWitnessValuesTest.java
package runar.lang.sdk;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

/**
 * R-6 — SDK consumer support for intent-intrinsic auto-injected witness
 * params ({@code _prevOutScript_<i>}, {@code _serialisedOutputs}).
 *
 * <p>Covers:
 * <ul>
 *   <li>filter: auto-injected witness params are NOT counted as user args
 *   <li>setters: {@code setPrevOutScript} / {@code setSerialisedOutputs}
 *   <li>errors: missing witness raises {@link WitnessValueMissingError}
 *   <li>wiring: witness bytes appear in the primary unlocking script in
 *       ABI order ({@code _prevOutScript_*} first, then
 *       {@code _serialisedOutputs})
 * </ul>
 *
 * <p>Tests use {@link RunarContract#prepareCall} and inspect the prepared
 * tx hex / unlock script directly — this lets us assert the witness wiring
 * without running the full deploy/call lifecycle through MockProvider.
 */
class IntentWitnessValuesTest {

    private static String paramJson(String name, String type) {
        return "{\"name\":\"" + name + "\",\"type\":\"" + type + "\"}";
    }

    /**
     * Build a stateless artifact with intent-witness params on its single
     * method. Stateless lets us exercise the filter + setter API without
     * needing the full stateful OP_PUSH_TX + funding pipeline.
     */
    private static RunarArtifact makeIntentArtifact(int[] prevOutInputs, boolean serialised) {
        StringBuilder params = new StringBuilder();
        params.append(paramJson("amount", "bigint"));
        for (int i : prevOutInputs) {
            if (params.length() > 0) params.append(",");
            params.append(paramJson("_prevOutScript_" + i, "ByteString"));
        }
        if (serialised) {
            if (params.length() > 0) params.append(",");
            params.append(paramJson("_serialisedOutputs", "ByteString"));
        }
        // Prepend amount comma if witness params followed it
        String paramsJson = paramJson("amount", "bigint");
        StringBuilder more = new StringBuilder();
        for (int i : prevOutInputs) {
            more.append(",").append(paramJson("_prevOutScript_" + i, "ByteString"));
        }
        if (serialised) {
            more.append(",").append(paramJson("_serialisedOutputs", "ByteString"));
        }
        paramsJson = paramsJson + more;

        String json =
            "{"
            + "\"version\":\"runar-v0.1.0\","
            + "\"compilerVersion\":\"0.1.0\","
            + "\"contractName\":\"IntentWitnessTest\","
            + "\"script\":\"51\","
            + "\"asm\":\"\","
            + "\"abi\":{"
                + "\"constructor\":{\"params\":[]},"
                + "\"methods\":[{\"name\":\"move\",\"isPublic\":true,\"params\":["
                + paramsJson + "]}]"
            + "},"
            + "\"buildTimestamp\":\"2026-05-18T00:00:00.000Z\""
            + "}";
        return RunarArtifact.fromJson(json);
    }

    private static RunarContract setupContract(RunarArtifact artifact) {
        UTXO contractUtxo = new UTXO("ab".repeat(32), 0, 10_000L, "51");
        RunarContract contract = new RunarContract(artifact, List.of());
        contract.setCurrentUtxo(contractUtxo);
        return contract;
    }

    // ---------------------------------------------------------------------
    // Filter: arg-count check excludes auto-injected witness params
    // ---------------------------------------------------------------------

    @Test
    void filterExcludesAutoInjectedWitnessParams() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{0, 1}, true);
        RunarContract contract = setupContract(artifact);
        contract.setPrevOutScript(0, "aa");
        contract.setPrevOutScript(1, "bb");
        contract.setSerialisedOutputs("cc");

        // 1 user arg passes (`amount`). Without the filter the SDK would
        // complain "expects 4 args" (amount + 2 prevOutScripts + serialisedOutputs).
        MockProvider provider = new MockProvider();
        PreparedCall pc = contract.prepareCall(
            "move",
            List.of(BigInteger.valueOf(123L)),
            null, provider, new MockSigner()
        );
        // Witness pushes should be in the prepared tx hex.
        String txHex = pc.txHex();
        assertTrue(txHex.contains("01aa"), "prevOut[0] push missing");
        assertTrue(txHex.contains("01bb"), "prevOut[1] push missing");
        assertTrue(txHex.contains("01cc"), "serialised push missing");
    }

    @Test
    void filterStillRejectsRealArgCountMismatches() {
        // The user-facing arg-count check lives in `call()` (the dispatch
        // entry point that splits into callStateless / callWithPushTx /
        // callTerminal). The `prepareCall` API doesn't enforce it
        // because it serves the external-signer flow where the caller
        // explicitly passes one arg slot per method param (sigs as null).
        // Verify the call-entry path rejects mismatched arg counts.
        RunarArtifact artifact = makeIntentArtifact(new int[]{0}, true);
        RunarContract contract = setupContract(artifact);
        // Supply witnesses so we don't hit WitnessValueMissingError first.
        contract.setPrevOutScript(0, "aa");
        contract.setSerialisedOutputs("bb");
        MockProvider provider = new MockProvider();
        MockSigner signer = new MockSigner();
        provider.addUtxo(signer.address(),
            new UTXO("ee".repeat(32), 0, 100_000L,
                ScriptUtils.buildP2PKHScript(signer.address())));

        // Pass 2 args when only `amount` is user-facing
        IllegalArgumentException err = assertThrows(
            IllegalArgumentException.class,
            () -> contract.call(
                "move", List.of(BigInteger.ONE, BigInteger.TWO), null, provider, signer
            )
        );
        assertTrue(err.getMessage().contains("expects 1 user args, got 2"),
            "expected arg count error: " + err.getMessage());
    }

    // ---------------------------------------------------------------------
    // Missing witness ⇒ typed WitnessValueMissingError
    // ---------------------------------------------------------------------

    @Test
    void missingPrevOutScriptRaisesTypedError() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{0}, false);
        RunarContract contract = setupContract(artifact);
        MockProvider provider = new MockProvider();

        WitnessValueMissingError err = assertThrows(
            WitnessValueMissingError.class,
            () -> contract.prepareCall(
                "move", List.of(BigInteger.ONE), null, provider, new MockSigner()
            )
        );
        assertEquals("_prevOutScript_0", err.paramName());
        assertEquals("move", err.methodName());
        assertEquals("IntentWitnessTest", err.contractName());
    }

    @Test
    void missingSerialisedOutputsRaisesTypedError() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{}, true);
        RunarContract contract = setupContract(artifact);
        MockProvider provider = new MockProvider();

        WitnessValueMissingError err = assertThrows(
            WitnessValueMissingError.class,
            () -> contract.prepareCall(
                "move", List.of(BigInteger.ONE), null, provider, new MockSigner()
            )
        );
        assertEquals("_serialisedOutputs", err.paramName());
    }

    // ---------------------------------------------------------------------
    // Wiring: witness bytes appear in the prepared unlocking script
    // ---------------------------------------------------------------------

    @Test
    void appendsMultiplePrevOutScriptsInAbiOrder() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{0, 1}, false);
        RunarContract contract = setupContract(artifact);
        contract.setPrevOutScript(0, "deadbeef");
        contract.setPrevOutScript(1, "cafebabe");
        MockProvider provider = new MockProvider();

        PreparedCall pc = contract.prepareCall(
            "move", List.of(BigInteger.ONE), null, provider, new MockSigner()
        );
        String txHex = pc.txHex();
        // PUSHDATA for 4 bytes = "04" + data
        String push0 = "04" + "deadbeef";
        String push1 = "04" + "cafebabe";
        int idx0 = txHex.indexOf(push0);
        int idx1 = txHex.indexOf(push1);
        assertNotEquals(-1, idx0, "witness 0 push missing in tx hex");
        assertTrue(idx1 > idx0,
            "witness 1 push must follow witness 0 (idx0=" + idx0 + ", idx1=" + idx1 + ")");
    }

    @Test
    void appendsPrevOutThenSerialisedInAbiOrder() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{0}, true);
        RunarContract contract = setupContract(artifact);
        contract.setPrevOutScript(0, "11223344");
        contract.setSerialisedOutputs("55667788");
        MockProvider provider = new MockProvider();

        PreparedCall pc = contract.prepareCall(
            "move", List.of(BigInteger.ONE), null, provider, new MockSigner()
        );
        String txHex = pc.txHex();
        int idxPrev = txHex.indexOf("0411223344");
        int idxSerial = txHex.indexOf("0455667788");
        assertNotEquals(-1, idxPrev, "prevOut push missing");
        assertTrue(idxSerial > idxPrev, "serialised push must follow prevOut push");
    }

    @Test
    void acceptsBytesViaConvenienceSetter() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{0}, false);
        RunarContract contract = setupContract(artifact);
        contract.setPrevOutScript(0, new byte[]{(byte) 0xab, (byte) 0xcd});
        MockProvider provider = new MockProvider();

        PreparedCall pc = contract.prepareCall(
            "move", List.of(BigInteger.ONE), null, provider, new MockSigner()
        );
        // 2-byte push = "02abcd"
        assertTrue(pc.txHex().contains("02abcd"), "2-byte witness push missing");
    }

    @Test
    void rejectsInvalidHex() {
        RunarArtifact artifact = makeIntentArtifact(new int[]{0}, false);
        RunarContract contract = new RunarContract(artifact, List.of());
        assertThrows(IllegalArgumentException.class,
            () -> contract.setPrevOutScript(0, "not-hex!"));
        assertThrows(IllegalArgumentException.class,
            () -> contract.setSerialisedOutputs("abc"));
    }

    private static void assertNotNull(Object value, String message) {
        if (value == null) throw new AssertionError(message);
    }
}

File: packages/runar-sdk/src/__tests__/intent-witness-values.test.ts
// ---------------------------------------------------------------------------
// runar-sdk/__tests__/intent-witness-values.test.ts
//
// R-6 — SDK consumer support for intent-intrinsic auto-injected witness params
// (`_prevOutScript_<i>`, `_serialisedOutputs`).
//
// Covers:
//   - filter: auto-injected witness params are NOT part of the user arg count
//   - setters: setPrevOutScript / setSerialisedOutputs store witness bytes
//   - errors: missing witness raises a typed WitnessValueMissingError
//   - wiring: witness bytes are appended to the primary unlocking script in
//     ABI order (`_prevOutScript_*` first, then `_serialisedOutputs`), AFTER
//     the BIP-143 preimage push and BEFORE the method selector
// ---------------------------------------------------------------------------

import { describe, it, expect } from 'vitest';
import { RunarContract } from '../contract.js';
import { MockProvider } from '../providers/mock.js';
import { LocalSigner } from '../signers/local.js';
import { WitnessValueMissingError } from '../errors.js';
import type { RunarArtifact, StateField } from 'runar-ir-schema';
import type { UTXO } from '../types.js';

const PRIV_KEY = '0000000000000000000000000000000000000000000000000000000000000001';

function makeFundingUtxo(satoshis: number, index = 0): UTXO {
  return {
    txid: 'aabbccdd'.repeat(8),
    outputIndex: index,
    satoshis,
    script: '76a914' + '00'.repeat(20) + '88ac',
  };
}

/**
 * Build a stateful artifact whose `move` method's ABI carries the
 * compiler-auto-injected continuation + intent-witness params. The on-chain
 * script is just OP_TRUE (`51`) so spending always succeeds — we only assert
 * on the unlocking-script BYTES, not on Bitcoin Script execution.
 */
function makeArtifactWithIntentWitness(
  extraWitnessParams: { prevOutInputs: number[]; serialised: boolean },
): RunarArtifact {
  const stateFields: StateField[] = [
    { name: 'count', type: 'bigint', index: 0 },
  ];

  const witnessParams = [
    ...extraWitnessParams.prevOutInputs.map((i) => ({
      name: `_prevOutScript_${i}`,
      type: 'ByteString',
    })),
    ...(extraWitnessParams.serialised
      ? [{ name: '_serialisedOutputs', type: 'ByteString' }]
      : []),
  ];

  return {
    version: 'runar-v0.1.0',
    compilerVersion: '0.1.0',
    contractName: 'IntentWitnessTest',
    asm: '',
    buildTimestamp: '2026-05-18T00:00:00.000Z',
    script: '51',
    abi: {
      constructor: { params: [{ name: 'count', type: 'bigint' }] },
      methods: [
        {
          name: 'move',
          isPublic: true,
          params: [
            // One ordinary user param
            { name: 'amount', type: 'bigint' },
            // Compiler-injected continuation params for stateful methods
            { name: '_changePKH', type: 'Ripemd160' },
            { name: '_changeAmount', type: 'bigint' },
            { name: '_newAmount', type: 'bigint' },
            { name: 'txPreimage', type: 'SigHashPreimage' },
            // Compiler-injected intent witness params (this is what R-6 covers)
            ...witnessParams,
          ],
        },
      ],
    },
    stateFields,
    codeSeparatorIndex: 0,
  };
}

describe('R-6 — intent-intrinsic witness values', () => {
  // -------------------------------------------------------------------------
  // Filter: arg-count check excludes _prevOutScript_* / _serialisedOutputs
  // -------------------------------------------------------------------------

  describe('arg-count filter', () => {
    it('does NOT count _prevOutScript_<i> or _serialisedOutputs in the user-facing arg count', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0, 1], serialised: true });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });

      // Supply witnesses so we don't trip WitnessValueMissingError — we're
      // asserting the filter passes the arg count check.
      contract.setPrevOutScript(0, 'aa');
      contract.setPrevOutScript(1, 'bb');
      contract.setSerialisedOutputs('cc');

      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      // User passes exactly 1 arg (just `amount`). Without the new filter
      // the SDK would have complained "expects 7 args, got 1" because the
      // ABI lists amount + 3 continuation params + 2 prevOutScripts + 1
      // serialisedOutputs (after filtering txPreimage out of the legacy
      // check). With the new filter only `amount` is user-facing.
      await contract.call('move', [123n], provider, signer, {
        newState: { count: 1n },
      });

      // No assertion error / no exception ⇒ filter accepted 1 user arg.
      expect(contract.state.count).toBe(1n);
    });

    it('still rejects user-arg count mismatches for non-auto-injected params', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0], serialised: true });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });

      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      // Pass 2 args when only `amount` is user-facing
      let err: unknown = null;
      try {
        await contract.call('move', [1n, 2n], provider, signer);
      } catch (e) {
        err = e;
      }
      expect(err).toBeInstanceOf(Error);
      expect((err as Error).message).toMatch(/expects 1 args, got 2/);
    });
  });

  // -------------------------------------------------------------------------
  // Missing witness ⇒ typed WitnessValueMissingError
  // -------------------------------------------------------------------------

  describe('missing witness raises WitnessValueMissingError', () => {
    it('throws when a `_prevOutScript_<i>` witness is not set', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0], serialised: false });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });
      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      let err: unknown = null;
      try {
        await contract.call('move', [1n], provider, signer);
      } catch (e) {
        err = e;
      }
      expect(err).toBeInstanceOf(WitnessValueMissingError);
      const wErr = err as WitnessValueMissingError;
      expect(wErr.paramName).toBe('_prevOutScript_0');
      expect(wErr.methodName).toBe('move');
      expect(wErr.contractName).toBe('IntentWitnessTest');
    });

    it('throws when `_serialisedOutputs` is not set', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [], serialised: true });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });
      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      let err: unknown = null;
      try {
        await contract.call('move', [1n], provider, signer);
      } catch (e) {
        err = e;
      }
      expect(err).toBeInstanceOf(WitnessValueMissingError);
      expect((err as WitnessValueMissingError).paramName).toBe('_serialisedOutputs');
    });
  });

  // -------------------------------------------------------------------------
  // Wiring: witness bytes appear in the broadcast unlocking script
  // -------------------------------------------------------------------------

  describe('witness bytes appear in primary unlocking script', () => {
    it('appends `_prevOutScript_*` witnesses (multi-input) in ABI order', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0, 1], serialised: false });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });
      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      const w0Hex = 'deadbeef';
      const w1Hex = 'cafebabe';
      contract.setPrevOutScript(0, w0Hex);
      contract.setPrevOutScript(1, w1Hex);

      await contract.call('move', [1n], provider, signer, { newState: { count: 1n } });

      const broadcastedTxs = provider.getBroadcastedTxs();
      expect(broadcastedTxs.length).toBe(2); // deploy + call
      const callTxHex = broadcastedTxs[1]!;

      // Find both witness pushes in ABI order. PUSHDATA for 4 bytes = `04` + data.
      const push0 = '04' + w0Hex;
      const push1 = '04' + w1Hex;
      const idx0 = callTxHex.indexOf(push0);
      const idx1 = callTxHex.indexOf(push1);
      expect(idx0).toBeGreaterThanOrEqual(0);
      expect(idx1).toBeGreaterThan(idx0); // _prevOutScript_1 follows _prevOutScript_0
    });

    it('appends both `_prevOutScript_<i>` and `_serialisedOutputs` with prevOuts FIRST', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0], serialised: true });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });
      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      const prevOutHex = '11223344';
      const serialisedHex = '55667788';
      contract.setPrevOutScript(0, prevOutHex);
      contract.setSerialisedOutputs(serialisedHex);

      await contract.call('move', [1n], provider, signer, { newState: { count: 1n } });

      const callTxHex = provider.getBroadcastedTxs()[1]!;
      const pushPrev = '04' + prevOutHex;
      const pushSerial = '04' + serialisedHex;
      const idxPrev = callTxHex.indexOf(pushPrev);
      const idxSerial = callTxHex.indexOf(pushSerial);
      expect(idxPrev).toBeGreaterThanOrEqual(0);
      expect(idxSerial).toBeGreaterThan(idxPrev);
    });

    it('accepts witness values as Uint8Array', async () => {
      const signer = new LocalSigner(PRIV_KEY);
      const address = await signer.getAddress();
      const provider = new MockProvider();
      provider.addUtxo(address, makeFundingUtxo(100_000));

      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0], serialised: false });
      const contract = new RunarContract(artifact, [0n]);
      await contract.deploy(provider, signer, { satoshis: 50_000 });
      provider.addUtxo(address, makeFundingUtxo(100_000, 1));

      contract.setPrevOutScript(0, new Uint8Array([0xab, 0xcd]));
      await contract.call('move', [1n], provider, signer, { newState: { count: 1n } });

      const callTxHex = provider.getBroadcastedTxs()[1]!;
      // 2-byte push: 02 + abcd
      expect(callTxHex.includes('02abcd')).toBe(true);
    });

    it('rejects hex inputs with invalid characters', () => {
      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0], serialised: false });
      const contract = new RunarContract(artifact, [0n]);
      expect(() => contract.setPrevOutScript(0, 'not-hex!')).toThrow();
    });

    it('rejects hex inputs with odd length', () => {
      const artifact = makeArtifactWithIntentWitness({ prevOutInputs: [0], serialised: false });
      const contract = new RunarContract(artifact, [0n]);
      expect(() => contract.setSerialisedOutputs('abc')).toThrow();
    });
  });
});
